src/Workspaces/Core/Portable/Rename/Renamer.RenameDocumentActionSet.cs

```diff
@@ -0,0 +1,120 @@
+                => UpdateSolutionAsync(solution, ApplicableActions, cancellationToken);
+
+            /// <summary>
+            /// Applies each <see cref="RenameDocumentAction"/> in order and returns the final solution. 
+            /// All actions must be contained in <see cref="ApplicableActions" />
+            /// </summary>
+            /// <remarks>
+            /// An empty action set is still allowed and will return a modified solution
+            /// that will update the document properties as appropriate.
+            /// </remarks>
+            public async Task<Solution> UpdateSolutionAsync(Solution solution, ImmutableArray<RenameDocumentAction> actions, CancellationToken cancellationToken)
+            {
+                if (actions.Any(a => !ApplicableActions.Contains(a)))
+                {
+                    throw new ArgumentException(WorkspacesResources.Cannot_apply_action_that_is_not_in_applicableactions);
+                }
+
+                // Prior to updating the solution it's possible the document id has changed between the time 
+                // the document action info was generated and when the solution update is applied. We
+                // do a best effort to still locate the document if the id has changed.
+                var document = GetDocument(solution);
+
+                // If the document was found in the solution then the current id will be durable across actions
+                // since we own the solution snapshot at this point. 
+                var documentId = document.Id;
+
+                // Make sure that the document name and folders are updated to what we expect them to be
+                solution = solution
+                    .WithDocumentName(documentId, _documentName)
+                    .WithDocumentFolders(documentId, _documentFolders);
+
+                // Apply each action individually. Order should not matter
+                foreach (var action in actions)
+                {
+                    document = solution.GetRequiredDocument(documentId);
+                    solution = await action.GetModifiedSolutionAsync(document, cancellationToken).ConfigureAwait(false);
+                }
+
+                return solution;
+            }
+
+            /// <summary>
+            /// Attempts to find the document in the solution. Tries by documentId first, but 
+            /// that's not always reliable between analysis and application of the rename actions
+            /// </summary>
+            private Document GetDocument(Solution solution)
+            {
+                // DocumentId is the best bet for finding a document,
+                // but it's possible the document was renamed or moved
+                // before actions were applied.
+                if (solution.ContainsDocument(_documentId))
+                {
+                    return solution.GetRequiredDocument(_documentId);
+                }
```

Comment:
If you want you can still use the old fashioned GetDocument that just returns null...

Reply:
I actually think this reads a little better than 

```csharp
var document = solution.GetDocument(_documentId);
if (document is object) 
{
    return document;
}
```

but don't feel strongly either way.